src/components/TodosContext/TodosContext.tsx

```diff
@@ -0,0 +1,104 @@
+import {
+  useEffect,
+  useReducer,
+  createContext,
+} from 'react';
+import { Todo } from '../types/Todo';
+
+export type Action =
+  | { type: 'add', payload: Todo }
+  | { type: 'remove', payload: number }
+  | { type: 'toggleComplete', payload: number }
+  | { type: 'toggleCompleteAll' }
+  | { type: 'toggleCompleted', payload: { id: number, newValue: string } }
+  | { type: 'edit', payload: { id: number, newValue: string } }
+  | { type: 'clearCompleted' };
+
+type State = Todo[];
+
+type Props = {
+  children: React.ReactNode,
+};
+
+const reducer = (state: State, action: Action): State => {
```

Comment:
why do you create reducer if you make context?
i think here you mixed to different approaches, it should be either context or redux